### PR TITLE
HOCS-3803: smc suspend feature

### PR DESCRIPTION
**Pull request by gary-newbolt**

### CODE REVIEW

File: server/lists/adapters/actions/suspensionsAdaptor.js [hunk 1]
```diff
@@ -0,0 +1,4 @@
+module.exports = (suspensions) => {
```
[the remark this answers is not in the document]
Comment (gary-newbolt) [reply server/lists/adapters/actions/suspensionsAdaptor.js:1]: Removed and replaced calling method to account for all actionData that does not need an adapter... so also removed the extensionsAdapter

File: server/lists/adapters/workstacks.js [hunk 2]
```diff
@@ -152,6 +152,24 @@ const getCorrespondentsNameByType = (correspondents, types) =>
         .map(correspondent => correspondent.fullname)
         .join(', ');
 
+/**
+ * Transforms stage.deadline value if case is/has been suspended.
+ *      Transform applied at this stage to allow Case Filter to be applied to filter on 'Suspended' and 'N/A'
+ * @param {object} stage - data for the stage.
+ */
+function resolveForSuspendedDeadlines(stage) {
```
[the remark this answers is not in the document]
Comment (gary-newbolt) [reply server/lists/adapters/workstacks.js:160]: I've abstracted the method out to a separate helper to handle both in `workstack.js` and `case-summary.js` and added a suite of tests for the 2 methods in the helper.

File: server/lists/adapters/case-summary.js [hunk 3]
```diff
@@ -90,6 +90,15 @@ const getPreviousCase = previousCase => {
...
 };
 
+const resolveDeadlineDisplay = summary => {
+    if (summary.suspended === 'true') {
+        return 'Suspended';
+    } else if (summary.caseDeadline === '9999-12-31') {
+        return 'N/A (case was previously suspended)';
+    }
+    return formatDate(summary.caseDeadline);
+};
```
[the remark this answers is not in the document]
Comment (gary-newbolt) [reply server/lists/adapters/case-summary.js:100]: I've abstracted the method out to a separate helper to handle both in `workstack.js` and `case-summary.js` and added a suite of tests for the 2 methods in the helper.

File: server/lists/adapters/workstacks.js [hunk 4]
```diff
@@ -218,6 +236,8 @@ const bindDisplayElements = fromStaticList => async (stage) => {
         stage.nextContributionDueDate = (stage.contributions === 'Overdue') ? 'Overdue' : formatDate(stage.dueContribution);
     }
 
+    resolveForSuspendedDeadlines(stage);
```
[the remark this answers is not in the document]
Comment (gary-newbolt) [reply server/lists/adapters/workstacks.js:239]: I've made the change as suggested. 

As for line 194, I've no idea of the difference. However a scan hocs-data suggests a very early config of workstage column defaults used this. I'm not sure if anything is still using this, `SMC` definitely isn't.

File: server/services/action.js [hunk 5]
```diff
@@ -328,15 +328,52 @@ const actions = {
...
+                                }
+                            };
+
+                            return handleActionSuccess(clientResponse, {}, form);
+                        }
+                        case actionTypes.UNSUSPEND_CASE: {
+
+                            const { caseActionId } = options;
+
+                            const response =  await caseworkService.delete(`/case/${caseId}/stage/${stageId}/actions/suspension/${caseActionId}`, headers);
```
[the remark this answers is not in the document]
Comment (gary-newbolt) [reply server/services/action.js:362]: updated to PUT request inline with `hocs-casework`

File: src/shared/pages/case.jsx [hunk 6]
```diff
@@ -35,11 +45,11 @@ class Case extends Component {
...
     shouldDisplaySidebar() {
-        if (this.hideSidebar === 'true') {
+        if (this.state.hideSidebar === 'true') {
             return false;
         }
 
-        if (this.hideSidebar === 'false') {
+        if (this.state.hideSidebar === 'false') {
             return true;
         }
```
[the remark this answers is not in the document]
Comment (gary-newbolt) [reply src/shared/pages/case.jsx:54]: I do not disagree... the best bit is that it didn't even work all the time until I change it to work from state instead. I'll add it to my raise a ticket list if that's ok?

File: src/shared/common/components/suspensions.jsx [hunk 7]
```diff
@@ -0,0 +1,85 @@
...
+        </div>
+    );
+};
+
+Suspensions.propTypes = {
+    page: PropTypes.object,
+    props: {
+        SUSPENSION: PropTypes.array
+    }
+};
```
[the remark this answers is not in the document]
Comment (gary-newbolt) [reply src/shared/common/components/suspensions.jsx]: @robertdeniszczyc2 I couldn't replicate the exception you found above. Regardless I tightened it up a bit. Hopefully it clears it up, but not sure what was wrong with it. If we can check on Monday I'd be grateful. Thanks

[the remark this answers is not in the document]
Comment (robertdeniszczyc2) [reply src/shared/common/components/suspensions.jsx]: Also can't replicate now

File: src/shared/pages/case.jsx (hunk 6 above)
[the remark this answers is not in the document]
Comment (robertdeniszczyc2) [reply src/shared/pages/case.jsx:54]: Did it ever work sans state, and I guess now with the change does the operations which previously worked under `this.hideSidebar` work with `state`?

Yeah we can raise a separate ticket for it, out of scope of this change

[the remark this answers is not in the document]
Comment (gary-newbolt) [reply src/shared/pages/case.jsx:54]: As far as I can tell... the change has not broken anything, however I haven't tested every possible form where this is applied. I think where the final return statement exists and references a prop that is used in the db schema's, and is subtly different `hasSidebar` I'm not sure if this worked either. 

I believe the problem was setting the value in this parent component cause refresh issues and the value did not propagate. Hence the change to store in state. It is possible anything that tried to use it in the past and code was left could be affected and now work, that might cause some regression test failures, which we should then discuss I feel.

I do agree another ticket should be raised to clean up the code and the functionality on it for sure.

[the remark this answers is not in the document]
Comment (robertdeniszczyc2) [reply src/shared/pages/case.jsx:54]: Ok cool yeah I agree let's put through regression

**Review by robertdeniszczyc2 — APPROVED** [review]

**Pull request merged**